assets/queries/terraform/azure/aks_rbac_disabled/query.rego

```diff
@@ -6,7 +6,7 @@ import data.generic.terraform as tf_lib
 CxPolicy[result] {
 	cluster := input.document[i].resource.azurerm_kubernetes_cluster[name]
 
-	not common_lib.valid_key(cluster, "role_based_access_control")
+	not common_lib.valid_key(cluster, "role_based_access_control_enabled")
```

Comment:
Hi, @rndmh3ro 🙂 

Thank you so much for contributing to KICS! 🚀 

Considering that we do not know if some user is using a version above `azurerm 3.0` and considering that [role_based_access_control_enabled](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#role_based_access_control_enabled) is set to true by default, I suggest the following:

```
package Cx

import data.generic.common as common_lib
import data.generic.terraform as tf_lib

CxPolicy[result] {
	cluster := input.document[i].resource.azurerm_kubernetes_cluster[name]

   # before azurerm 3.0
	rbac := cluster.role_based_access_control

	rbac.enabled != true

	result := {
		"documentId": input.document[i].id,
		"resourceType": "azurerm_kubernetes_cluster",
		"resourceName": tf_lib.get_resource_name(cluster, name),
		"searchKey": sprintf("azurerm_kubernetes_cluster[%s].role_based_access_control.enabled", [name]),
		"searchLine": common_lib.build_search_line(["resource","azurerm_kubernetes_cluster", name ,"role_based_access_control", "enabled"], []),
		"issueType": "IncorrectValue",
		"keyExpectedValue": sprintf("'azurerm_kubernetes_cluster[%s].role_based_access_control.enabled' should be set to true", [name]),
		"keyActualValue": sprintf("'azurerm_kubernetes_cluster[%s].role_based_access_control.enabled' is not set to true", [name]),
		"remediation": json.marshal({
			"before": "false",
			"after": "true"
		}),
		"remediationType": "replacement",
	}
}

CxPolicy[result] {
	cluster := input.document[i].resource.azurerm_kubernetes_cluster[name]

   # after azurerm 3.0
	rbac := cluster.role_based_access_control_enabled != true

	result := {
		"documentId": input.document[i].id,
		"resourceType": "azurerm_kubernetes_cluster",
		"resourceName": tf_lib.get_resource_name(cluster, name),
		"searchKey": sprintf("azurerm_kubernetes_cluster[%s].role_based_access_control_enabled", [name]),
		"searchLine": common_lib.build_search_line(["resource","azurerm_kubernetes_cluster", name ,"role_based_access_control_enabled"], []),
		"issueType": "IncorrectValue",
		"keyExpectedValue": sprintf("'azurerm_kubernetes_cluster[%s].role_based_access_control_enabled' should be set to true", [name]),
		"keyActualValue": sprintf("'azurerm_kubernetes_cluster[%s].role_based_access_control_enabled' is not set to true", [name]),
		"remediation": json.marshal({
			"before": "false",
			"after": "true"
		}),
		"remediationType": "replacement",
	}
}
```

If you agree, I also suggest keeping the "old samples" and adding your new ones

Reply:
How does this work? 


I guess it has something to do with the `should`:

```
-               "keyExpectedValue": sprintf("'azurerm_kubernetes_cluster[%s].role_based_access_control_enabled' is set to true", [name]),
+               "keyExpectedValue": sprintf("'azurerm_kubernetes_cluster[%s].role_based_access_control_enabled' should be set to true", [name]),
```

Does the `should` indicate that only one of the policies in the file has to match?

---

If I use your code, I get the following result:

```
AKS RBAC Disabled, Severity: MEDIUM, Results: 1
Description: Azure Container Service (AKS) instance should have role-based access control (RBAC) enabled
Platform: Terraform

        [1]: terraform/aks.tf:29

                028:   kubernetes_version                = "1.22.6"
                029:   role_based_access_control_enabled = true
                030:   tags                              = local.tags
```

However with this line changed it works:
```
-    rbac := cluster.role_based_access_control_enabled != true
+    cluster.role_based_access_control_enabled != true
```

If you're fine with my change, I'll update the PR.

Reply:
> How does this work?
> 
> I guess it has something to do with the `should`:
> 
> ```
> -               "keyExpectedValue": sprintf("'azurerm_kubernetes_cluster[%s].role_based_access_control_enabled' is set to true", [name]),
> +               "keyExpectedValue": sprintf("'azurerm_kubernetes_cluster[%s].role_based_access_control_enabled' should be set to true", [name]),
> ```
> 
> Does the `should` indicate that only one of the policies in the file has to match?

No. I suggest changing `"keyExpectedValue": sprintf("'azurerm_kubernetes_cluster[%s].role_based_access_control_enabled' is set to true", [name]),"` to `"keyExpectedValue": sprintf("'azurerm_kubernetes_cluster[%s].role_based_access_control_enabled' should be set to true", [name]),"` because we are standardizing the `keyExcepectedValue` in order to suggest "the recommended change". It is not related to query approach.


> 
> If I use your code, I get the following result:
> 
> ```
> AKS RBAC Disabled, Severity: MEDIUM, Results: 1
> Description: Azure Container Service (AKS) instance should have role-based access control (RBAC) enabled
> Platform: Terraform
> 
>         [1]: terraform/aks.tf:29
> 
>                 028:   kubernetes_version                = "1.22.6"
>                 029:   role_based_access_control_enabled = true
>                 030:   tags                              = local.tags
> ```
> 
> However with this line changed it works:
> 
> ```
> -    rbac := cluster.role_based_access_control_enabled != true
> +    cluster.role_based_access_control_enabled != true
> ```
> 
> If you're fine with my change, I'll update the PR.

Sorry! My bad! I copied and pasted the first policy, and I forgot to delete `rbac:=`. Yes, of course, thank you so much!

Reply:
I'll do that on monday, thanks for reviewing!

Reply:
The first policy will only return a result if `role_based_access_control` is defined and `role_based_access_control.enabled` is defined but not set to true.

The second policy will only return a result if `role_based_access_control_enabled` is defined but not set to true.

Reply:
Thank you so much, @rndmh3ro! Have a nice weekend 😊

Reply:
I need some help with fixing the tests, please.
I now have three queries: the two previous ones and the new one. However the tests fail:
```
--- FAIL: TestQueries (0.07s)
    --- FAIL: TestQueries/terraform/azure/aks_rbac_disabled_positive (0.05s)
        queries_test.go:343:
                Error Trace:    /home/segu/dev/kics/test/queries_test.go:343
                                                        /home/segu/dev/kics/test/queries_test.go:260
                                                        /home/segu/dev/kics/test/queries_test.go:163
                Error:          "[{%!s(int=0) test_scan cac0beafd0b2dc2741a931699bec84dd300147e9033bbc730ac1c4b91eb5f234 354d042a-28bd-485d-b682-1e85f66b8d62 ../assets/queries/terraform/azure/aks_rbac_disabled/test/positive.tf 86f92117-eed8-4614-9c6c-b26da20ff37f AKS RBAC Disabled https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#role_based_access_control Access Control Azure Container Service (AKS) instance should have role-based access control (RBAC) enabled d27166a7 Terraform MEDIUM %!s(int=1) [] azurerm_kubernetes_cluster example-aks1 MissingAttribute azurerm_kubernetes_cluster[positive1] %!s(int=1)  'azurerm_kubernetes_cluster[positive1].role_based_access_control' is defined and not null 'azurerm_kubernetes_cluster[positive1].role_based_access_control' is undefined or null %!s(*string=<nil>) {"category":"Access Control","cloudProvider":"azure","descriptionID":"d27166a7","descriptionText":"Azure Container Service (AKS) instance should have role-based access control (RBAC) enabled","descriptionUrl":"https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#role_based_access_control","documentId":"354d042a-28bd-485d-b682-1e85f66b8d62","id":"86f92117-eed8-4614-9c6c-b26da20ff37f","issueType":"MissingAttribute","keyActualValue":"'azurerm_kubernetes_cluster[positive1].role_based_access_control' is undefined or null","keyExpectedValue":"'azurerm_kubernetes_cluster[positive1].role_based_access_control' is defined and not null","platform":"Terraform","queryName":"AKS RBAC Disabled","remediation":"role_based_access_control {\n\t\tenabled = true\n\t}","remediationType":"addition","resourceName":"example-aks1","resourceType":"azurerm_kubernetes_cluster","searchKey":"azurerm_kubernetes_cluster[positive1]","searchLine":["resource","azurerm_kubernetes_cluster","positive1"],"severity":"MEDIUM"} azure role_based_access_control {
                                                enabled = true
                                        } addition} {%!s(int=0) test_scan 0d15309b68491031f4893783fc208eacad31a5bbc92778469ba2cacc50a8479e 354d042a-28bd-485d-b682-1e85f66b8d62 ../assets/queries/terraform/azure/aks_rbac_disabled/test/positive.tf 86f92117-eed8-4614-9c6c-b26da20ff37f AKS RBAC Disabled https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#role_based_access_control Access Control Azure Container Service (AKS) instance should have role-based access control (RBAC) enabled d27166a7 Terraform MEDIUM %!s(int=7) [{%!s(int=7)   role_based_access_control_enabled = false}] azurerm_kubernetes_cluster example-aks1 IncorrectValue azurerm_kubernetes_cluster[positive1].role_based_access_control_enabled %!s(int=7)  'azurerm_kubernetes_cluster[positive1].role_based_access_control_enabled' should be set to true 'azurerm_kubernetes_cluster[positive1].role_based_access_control_enabled' is not set to true %!s(*string=<nil>) {"category":"Access Control","cloudProvider":"azure","descriptionID":"d27166a7","descriptionText":"Azure Container Service (AKS) instance should have role-based access control (RBAC) enabled","descriptionUrl":"https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#role_based_access_control","documentId":"354d042a-28bd-485d-b682-1e85f66b8d62","id":"86f92117-eed8-4614-9c6c-b26da20ff37f","issueType":"IncorrectValue","keyActualValue":"'azurerm_kubernetes_cluster[positive1].role_based_access_control_enabled' is not set to true","keyExpectedValue":"'azurerm_kubernetes_cluster[positive1].role_based_access_control_enabled' should be set to true","platform":"Terraform","queryName":"AKS RBAC Disabled","remediation":"{\"after\":\"true\",\"before\":\"false\"}","remediationType":"replacement","resourceName":"example-aks1","resourceType":"azurerm_kubernetes_cluster","searchKey":"azurerm_kubernetes_cluster[positive1].role_based_access_control_enabled","searchLine":["resource","azurerm_kubernetes_cluster","positive1","role_based_access_control_enabled"],"severity":"MEDIUM"} azure {"after":"true","before":"false"} replacement} {%!s(int=0) test_scan 3786cb02aca05062c5369ebc20ca4cbbda1a1404324fe6269fb142c11015a8f1 354d042a-28bd-485d-b682-1e85f66b8d62 ../assets/queries/terraform/azure/aks_rbac_disabled/test/positive.tf 86f92117-eed8-4614-9c6c-b26da20ff37f AKS RBAC Disabled https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#role_based_access_control Access Control Azure Container Service (AKS) instance should have role-based access control (RBAC) enabled d27166a7 Terraform MEDIUM %!s(int=35) [{%!s(int=35)     enabled = false}] azurerm_kubernetes_cluster example-aks2 IncorrectValue azurerm_kubernetes_cluster[positive2].role_based_access_control.enabled %!s(int=35)  'azurerm_kubernetes_cluster[positive2].role_based_access_control.enabled' should be set to true 'azurerm_kubernetes_cluster[positive2].role_based_access_control.enabled' is not set to true %!s(*string=<nil>) {"category":"Access Control","cloudProvider":"azure","descriptionID":"d27166a7","descriptionText":"Azure Container Service (AKS) instance should have role-based access control (RBAC) enabled","descriptionUrl":"https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#role_based_access_control","documentId":"354d042a-28bd-485d-b682-1e85f66b8d62","id":"86f92117-eed8-4614-9c6c-b26da20ff37f","issueType":"IncorrectValue","keyActualValue":"'azurerm_kubernetes_cluster[positive2].role_based_access_control.enabled' is not set to true","keyExpectedValue":"'azurerm_kubernetes_cluster[positive2].role_based_access_control.enabled' should be set to true","platform":"Terraform","queryName":"AKS RBAC Disabled","remediation":"{\"after\":\"true\",\"before\":\"false\"}","remediationType":"replacement","resourceName":"example-aks2","resourceType":"azurerm_kubernetes_cluster","searchKey":"azurerm_kubernetes_cluster[positive2].role_based_access_control.enabled","searchLine":["resource","azurerm_kubernetes_cluster","positive2","role_based_access_control","enabled"],"severity":"MEDIUM"} azure {"after":"true","before":"false"} replacement} {%!s(int=0) test_scan 9ad50020524c7c59465dff5c89718bb07e9a756a169fa49b8ebe79fd102b5dd6 10437a7e-b6b0-4a02-9ce6-1058e30355c4 ../assets/queries/terraform/azure/aks_rbac_disabled/test/positive2.tf 86f92117-eed8-4614-9c6c-b26da20ff37f AKS RBAC Disabled https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#role_based_access_control Access Control Azure Container Service (AKS) instance should have role-based access control (RBAC) enabled d27166a7 Terraform MEDIUM %!s(int=1) [] azurerm_kubernetes_cluster example-aks2 MissingAttribute azurerm_kubernetes_cluster[positive3] %!s(int=1)  'azurerm_kubernetes_cluster[positive3].role_based_access_control' is defined and not null 'azurerm_kubernetes_cluster[positive3].role_based_access_control' is undefined or null %!s(*string=<nil>) {"category":"Access Control","cloudProvider":"azure","descriptionID":"d27166a7","descriptionText":"Azure Container Service (AKS) instance should have role-based access control (RBAC) enabled","descriptionUrl":"https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#role_based_access_control","documentId":"10437a7e-b6b0-4a02-9ce6-1058e30355c4","id":"86f92117-eed8-4614-9c6c-b26da20ff37f","issueType":"MissingAttribute","keyActualValue":"'azurerm_kubernetes_cluster[positive3].role_based_access_control' is undefined or null","keyExpectedValue":"'azurerm_kubernetes_cluster[positive3].role_based_access_control' is defined and not null","platform":"Terraform","queryName":"AKS RBAC Disabled","remediation":"role_based_access_control {\n\t\tenabled = true\n\t}","remediationType":"addition","resourceName":"example-aks2","resourceType":"azurerm_kubernetes_cluster","searchKey":"azurerm_kubernetes_cluster[positive3]","searchLine":["resource","azurerm_kubernetes_cluster","positive3"],"severity":"MEDIUM"} azure role_based_access_control {
                                                enabled = true
                                        } addition}]" should have 3 item(s), but has 4
                Test:           TestQueries/terraform/azure/aks_rbac_disabled_positive
                Messages:       Count of actual issues and expected vulnerabilities doesn't match
                                 --
                                ../assets/queries/terraform/azure/aks_rbac_disabled/test/positive.tf:1,
                                ../assets/queries/terraform/azure/aks_rbac_disabled/test/positive.tf:35,
                                ../assets/queries/terraform/azure/aks_rbac_disabled/test/positive2.tf:1
FAIL
FAIL    github.com/Checkmarx/kics/test  0.119s
FAIL
```

Probably because the old one (with `not common_lib.valid_key(cluster, "role_based_access_control")`) now works on two test-cases?

Reply:
Yes. We will need to delete the first policy, @rndmh3ro